### DISCUSSION TITLE
Looking for docs / guidance on Blazor Web View communications with app code (and vice-versa)

### DISCUSSION THREAD

**Discussion by domingoladron** [dotnet/maui#5087] [General]
So I've got a baseline Blazor Maui app running and that's sweet, but what's still patently unclear is HOW you bridge the logical universe between a BlazorWebView and the code outside of the WebView.

So, for example, I have a simple Blazor app that allows for editing of files, but then I want the content of those files to be fed back to the desktop app so that can write this data to a directory / file.

Are there any such docs / guidance around?  Understanding this communications paradigm is pretty critical to the successful development of an app.

I am assuming there is some kind of eventing system in the Blazor side which allows you to call out from the Blazor code and then have say the Xaml logic in the desktop app to register an event handler to listen for such messages.

**Comment by domingoladron** [comment]
Found the answer.  Cheers

**Reply by Dresel** [reply]
Mind sharing?

**Reply by domingoladron** [reply]
@Dresel 

Sorry for the delay in responding. I wanted to confirm what I'm about to say so I didn't lead anyone down any primrose paths.  I'm still not 100% sure this is correct, mostly because the docs on WebView2, BlazorWebView and Maui are currently quite sketchy and sparse.

As far as I can determine, in a Maui app I didn't really need a communications channel between the Xaml and the BlazorWebView, simply because the BlazorWebView is hooked directly into the .NET Core Framework, so you can do all the things inside your Blazor code you'd normally do inside a normal .NET Core desktop application.  For example, I'm building an app which saves files based on user input, and the Blazor code inside the BlazorWebView just writes files to disk as any .NET desktop app would do so.  The BlazorWebView appears to be capable of talking right to .NET functions which for a Blazor WASM app are totally disallowed.

I noted this comment on the [BlazorWebView documentation page](https://docs.microsoft.com/en-us/dotnet/maui/user-interface/controls/blazorwebview):

`The [Razor components](https://docs.microsoft.com/en-us/aspnet/core/blazor/components/) run natively in the .NET process and render web UI to an embedded web view control.`

So, unlike the WebView2 component, which seems to obey the standard browser privacy and security rules (where communications takes place via Javascript -> C# eventing), with the BlazorWebView inside a Maui app, you can wire up whatever kind of communications you like between the two sides, simply because they are -- in essence -- the same app.

This freed me up as I no longer needed to bother trying to create a communication channel into / out of the BlazorWebView as you have to do with the WebView2.    Since the WPF logic and the Blazor logic appear to be using the same classes / configuration / IoC and so on, it became just a matter of changing my thinking to have my Blazor components raise shared Events which my WPF code is listening for and then acting accordingly.

Hope this helps.

**Reply by LeviSTA** [reply]
How did you end up doing the communication?
